jina/peapods/networking.py

```diff
@@ -1,8 +1,341 @@
+        return connection_pool
+
+    def start(self):
+        """
+        Starts the connection pool
+        """
+        pass
+
+    def close(self):
+        """
+        Closes the connection pool
+        """
+        self._connections.clear()
+
+    @abstractmethod
+    def _send_message(self, msg: Message, connection):
+        ...
+
+    @abstractmethod
+    def _create_connection(self, target):
+        ...
+
+
+class GrpcConnectionPool(ConnectionPool):
+    """
+    GrpcConnectionPool which uses gRPC as the communication mechanism
+    """
+
+    def _send_message(self, msg: Message, connection):
+        # this wraps the awaitable object from grpc as a coroutine so it can be used as a task
+        # the grpc call function is not a coroutine but some _AioCall
+        async def task_wrapper(new_message, stub):
+            await stub.Call(new_message)
+
+        return asyncio.create_task(task_wrapper(msg, connection))
+
+    def _create_connection(self, target):
+        self._logger.debug(f'create connection to {target}')
+        channel = grpc.aio.insecure_channel(
+            target,
+            options=[
+                ('grpc.max_send_message_length', -1),
+                ('grpc.max_receive_message_length', -1),
+            ],
+        )
+
+        return jina_pb2_grpc.JinaDataRequestRPCStub(channel)
+
+
+class K8sGrpcConnectionPool(GrpcConnectionPool):
+    """
+    Manages grpc connections to replicas in a K8s deployment.
+
+    :param namespace: K8s namespace to operate in
+    :param client: K8s client
+    :param logger: the logger to use
+    """
+
+    def __init__(
+        self,
+        namespace: str,
+        client,
+        logger: JinaLogger = None,
+    ):
+        super().__init__(logger=logger, on_demand_connection=False)
+
+        self._namespace = namespace
+        self._deployment_clusteraddresses = {}
+        self._k8s_client = client
+        self._k8s_event_queue = asyncio.Queue()
+        self.enabled = False
+
+        self._fetch_initial_state()
+
+        from kubernetes import watch
+
+        self._api_watch = watch.Watch()
+
+        self.update_thread = Thread(target=self.run, daemon=True)
+
+    def _fetch_initial_state(self):
+        namespaced_pods = self._k8s_client.list_namespaced_pod(self._namespace)
+        for item in namespaced_pods.items:
+            self._process_item(item)
+
+    def start(self):
+        """
+        Subscribe to the K8s API and watch for changes in Pods
+        """
+        self._loop = get_or_reuse_loop()
+        self._process_events_task = asyncio.create_task(self._process_events())
+        self.update_thread.start()
+
+    async def _process_events(self):
+        while self.enabled:
+            event = await self._k8s_event_queue.get()
+            self._process_item(event)
+
+    def run(self):
+        """
+        Subscribes on MODIFIED events from list_namespaced_pod AK8s PI
+        """
+
+        self.enabled = True
```

Comment:
Why is this important? I think that the correct way to do things would be to instantiate the `_k8s_client` locally in the `stack` of being run better than being a shared resource between the process and its thread

Reply:
Same for `loop` Is not clear to me which are the `shared` resources with the `thread`. And are they safe to be accessed without `locking`?

Reply:
Lets clear up the shared client in a follow up issue.
The loop I need to do 
`asyncio.run_coroutine_threadsafe(self._k8s_event_queue.put(event['object']), self._loop )`

I use an asyncio.Queue for communication between the Threads and this Queue is not threadsafe so I use the loop here to access the queue in a safe (but ugly) manner